client/components/timezone/index.jsx

```diff
@@ -15,9 +16,32 @@ import QueryTimezones from 'components/data/query-timezones';
 import getRawOffsets from 'state/selectors/get-raw-offsets';
 import getTimezones from 'state/selectors/get-timezones';
 
+function maybeFindLinkedTimezone( timezone, timezones ) {
+	if ( ! timezones || timezone in timezones ) {
+		return timezone;
+	}
+	// The (guessed) timezone could not be found, so let's see if a linked timezone for this exists.
+	// Example: we don't have Asia/Calcutta but Asia/Kalkota, so we'll need to choose that.
+	timezone = reduce(
+		pickBy( flattenDepth( timezones, 2 ), isArray ),
```

Comment:
This kind of code is very difficult to follow if I don't know what the structure of the `timezones` is. Can you add some explanatory comments?

Reply:
The `pickBy( isArray )` is to filter the continent names from the doubly flattened list that looks like this?
```js
 [ 'Africa', [ 'Africa/A', 'A' ], [ 'Africa/B', 'B' ], 'America', [ 'America/C' ], ... ]
```
A better way to extract just the timezones is:
```js
timezones.flatMap( ( [ continent, zones ] ) => zones );
```

Reply:
Another way to do this would be to pull the data from https://data.iana.org/time-zones/tzdb-2018h/backward (or moment.tz._links) into a hash or array of "synonyms" and store it in a static file. This way the data structure will be more easily accessible, it'll be closer to what's needed for the task, and you won't be depending on an internal property of `moment`.

Regardless of the approach it looks like this would also be a good piece of logic to write a couple of tests for.

(I got notified from the comment on https://github.com/Automattic/wp-calypso/pull/23226 in case you're wondering.)

Reply:
> into a hash or array of "synonyms" and store it in a file bundled with Calypso.

This is a nice idea and would certainly make the code more readable since the data structure is  visible. My only concern is that if we store the data in Calypso, then we need to take care of updating the data. What if some new city gets renamed and added to the list? `moment.tz._links` would have the updated values but the local copy wouldn't. This is going to be a rare event though so certainly not a big concern.

Reply:
👋  @nylen thanks for chiming in :) 
I was considering that but wanted to do it the "right way" first. But I guess the list is really static and it's probably an easier approach to just hardcode it, since we actually need to go from "new" to "old" (see https://core.trac.wordpress.org/ticket/26656) and can therefore manually curate the list.